### PR TITLE
typing: Improve FixtureDefinition and FixtureDef

### PR DESCRIPTION
**Pull request by nicoddemus**
Follow up to #12473, built on top of it, not ready to merge.

### CODE REVIEW

**Review by Tusenka — APPROVED** [review]

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus can you open a pr against the branch @Glyphack has from your branch and link it, so the change can be seen in relation to his pr

File: src/_pytest/fixtures.py [hunk 1]
```diff
@@ -124,7 +130,7 @@ def pytest_sessionstart(session: Session) -> None:
 
 def get_scope_package(
     node: nodes.Item,
-    fixturedef: FixtureDef[object],
+    fixturedef: FixtureDef[Any, object],
```
Comment (Tusenka) [inline src/_pytest/fixtures.py:133]: what does Any represent here?

File: src/_pytest/fixtures.py [hunk 2]
```diff
@@ -1223,19 +1207,21 @@ class FixtureFunctionMarker:
...
-    def __call__(self, function: FixtureFunction) -> FixtureFunction:
+    def __call__(
+        self, function: Callable[FixtureParams, FixtureValue]
+    ) -> FixtureFunctionDefinition[FixtureParams, FixtureValue]:
         if inspect.isclass(function):
             raise ValueError("class fixtures not supported (maybe in the future)")
 
-        if getattr(function, "_pytestfixturefunction", False):
+        if isinstance(function, FixtureFunctionDefinition):
             raise ValueError(
```
Comment (Tusenka) [inline src/_pytest/fixtures.py:1217]: in which cases this error could be raised

File: src/_pytest/fixtures.py (hunk 1 above)
Comment (RonnyPfannschmidt) [reply src/_pytest/fixtures.py:133]: It's for filling the new generic parameter for the arguments

**Comment by nicoddemus** [comment]
Superseded by #13036.

**Pull request closed without merging**